src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs

```diff
@@ -3,311 +3,819 @@
+            {
+                _writer.Write(CacheFormatSignature);
+                _writer.Write(CacheFormatVersion);
+
+                byte[] hash = _task.HashSettings();
+                _writer.Write(_task.HashSettings());
+                _metadataStringTablePosition = WritePlaceholder();
+            }
+
+            private void WriteItemGroups()
+            {
+                // NOTE: Order (alphabetical by group name) must match reader.
+                WriteItemGroup(WriteAnalyzers);
+                WriteItemGroup(WriteCompileTimeAssemblies);
+                WriteItemGroup(WriteContentFilesToPreprocess);
+                WriteItemGroup(WriteFrameworkAssemblies);
+                WriteItemGroup(WriteLogMessages);
+                WriteItemGroup(WriteNativeLibraries);
+                WriteItemGroup(WriteResourceAssemblies);
+                WriteItemGroup(WriteRuntimeAssemblies);
+                WriteItemGroup(WriteRuntimeTargets);
+                WriteItemGroup(WriteTransitiveProjectReferences);
+            }
+
+            private void WriteMetadataStringTable()
+            {
+                int savedPosition = Position;
+
+                _writer.Write(_metadataStrings.Count);
+
+                foreach (var s in _metadataStrings)
+                {
+                    _writer.Write(s);
+                }
+
+                WriteToPlaceholder(_metadataStringTablePosition, savedPosition);
+            }
+
+            private int Position
+            {
+                get => checked((int)_writer.BaseStream.Position);
+                set => _writer.BaseStream.Position = value;
+            }
+
+            private struct Placeholder
+            {
+                public readonly int Position;
+                public Placeholder(int position) { Position = position; }
+            }
+
+            private Placeholder WritePlaceholder()
+            {
+                var placeholder = new Placeholder(Position);
+                _writer.Write(int.MinValue);
+                return placeholder;
+            }
+
+            private void WriteToPlaceholder(Placeholder placeholder, int value)
+            {
+                int savedPosition = Position;
+                Position = placeholder.Position;
+                _writer.Write(value);
+                Position = savedPosition;
+            }
+
+            private void WriteAnalyzers()
+            {
+                Dictionary<string, LockFileTargetLibrary> targetLibraries = null;
+
+                foreach (var library in _lockFile.Libraries)
+                {
+                    if (!library.IsPackage())
+                    {
+                        continue;
+                    }
+
+                    foreach (var file in library.Files)
                     {
-                        if (!asset.IsPlaceholderFile() && (filter == null || filter(asset)))
+                        if (!NuGetUtils.IsApplicableAnalyzer(file, _task.ProjectLanguage))
```

Comment:
This class combines two things: Logic for how we interpret the contents of the lock file, and the format we use to cache it.  It seems like it would be cleaner to separate these two things into two different classes.  Is this possible, or do they need to be mixed in order to maintain the perf gains from the cache?

Reply:
They are mixed but the methods should be reasonably separated enough. This was indeed for perf so that we don't end up with any intermediate data to allocate when going from assets.json -> assets.cache.

Reply:
This is covered by #2048. See https://github.com/dotnet/sdk/pull/2020#discussion_r174928582